File: http-server-tomcat/src/test/java/io/micronaut/servlet/tomcat/websocket/ChatClientWebSocket.java
package io.micronaut.servlet.tomcat.websocket;

import io.micronaut.context.annotation.Requires;
import io.micronaut.http.HttpRequest;
import io.micronaut.websocket.CloseReason;
import io.micronaut.websocket.WebSocketSession;
import io.micronaut.websocket.annotation.ClientWebSocket;
import io.micronaut.websocket.annotation.OnClose;
import io.micronaut.websocket.annotation.OnMessage;
import io.micronaut.websocket.annotation.OnOpen;

import java.util.Collection;
import java.util.concurrent.ConcurrentLinkedQueue;

@Requires(property = "spec.name", value = "TomcatWebSocketSpec")
@ClientWebSocket("/chat/{topic}/{username}")
public abstract class ChatClientWebSocket implements AutoCloseable {

    private WebSocketSession session;
    private HttpRequest<?> request;
    private String topic;
    private String username;
    private CloseReason closeReason;
    private final Collection<String> replies = new ConcurrentLinkedQueue<>();

    @OnOpen
    public void onOpen(String topic, String username, WebSocketSession session, HttpRequest<?> request) {
        this.topic = topic;
        this.username = username;
        this.session = session;
        this.request = request;
    }

    @OnMessage
    public void onMessage(String message) {
        replies.add(message);
    }

    @OnClose
    public void onClose(CloseReason closeReason) {
        this.closeReason = closeReason;
    }

    public abstract void send(String message);

    public String getTopic() {
        return topic;
    }

    public String getUsername() {
        return username;
    }

    public Collection<String> getReplies() {
        return replies;
    }

    public WebSocketSession getSession() {
        return session;
    }

    public HttpRequest<?> getRequest() {
        return request;
    }

    public CloseReason getCloseReason() {
        return closeReason;
    }
}

File: test-suite-python/src/test/python/micronaut/servlet/docs/websocket/ChatClientWebSocket.py
from abc import ABC, abstractmethod

from java.lang import AutoCloseable
from java.util import Collection
from java.util.concurrent import ConcurrentLinkedQueue
from micronaut.context.annotation import Requires
from micronaut.websocket.annotation import ClientWebSocket, OnMessage


@Requires(property="spec.name", value="ChatWebSocketSpec")
# tag::class[]
@ClientWebSocket("/ws/chat/{topic}/{username}")  # <1>
class ChatClientWebSocket(ABC, AutoCloseable):  # <2>

    def __init__(self):
        self.replies = ConcurrentLinkedQueue()

    @OnMessage
    def on_message(self, message: str) -> None:
        self.replies.add(message)  # <3>

    @abstractmethod
    def send(self, message: str) -> None:  # <4>
        ...

    def get_replies(self) -> Collection[str]:
        return self.replies
# end::class[]
